client/pov-recovery/src/lib.rs

```diff
@@ -343,13 +344,55 @@ where
 		self.import_block(block).await;
 	}
 
+	async fn chain_recovery(&mut self, mut hash: Block::Hash) {
+		let mut to_recover = Vec::new();
+		let do_recover = loop {
+			match self.parachain_client.block_status(&BlockId::Hash(hash)) {
+				Ok(BlockStatus::Unknown) => to_recover.push(hash),
+				Ok(_) => break true,
+				Err(e) => {
+					tracing::error!(
+						target: LOG_TARGET,
+						error = ?e,
+						block_hash = ?hash,
+						"Failed to get block status",
+					);
+					break false
+				},
+			}
+
+			hash = match self.pending_candidates.get(&hash) {
+				Some(p) => p.parent_hash,
+				None => {
+					tracing::error!(
+						target: LOG_TARGET,
+						block_hash = ?hash,
+						"Cound not recover, block has never been announced as a candidate"
+					);
+					break false
+				},
+			}
+		};
+
+		if do_recover {
+			for hash in to_recover.iter().rev() {
+				tracing::debug!(
+					target: LOG_TARGET,
+					block_hash = ?hash,
+					"Starting eager recovery",
+				);
+				self.recover_candidate(*hash).await;
```

Comment:
Why are we not needing the delay here? This could potentially lead to many recoveries at the same time.

Reply:
The delay has not been added because we here we want to trigger the recovery immediately for all the missing blocks. Do you think that is better to add a random delay anyway? E.g. to spread the recovery in a bigger timespan?

Reply:
Per this comment, the delay prevents spamming of the relay chain. 
https://github.com/skunert/cumulus/blob/master/client/pov-recovery/src/lib.rs#L209

As I understand, your level monitor behaves deterministically on all collators. So if there were a problem with the relay chain and all collators would hit the level limit, the monitor would delete the same branches on every collator, right? This also means that when a new best block finally comes in, all collators would request recovery of the same blocks simultaneously (given that it was deleted by the level monitor). 

Mabe @bkchr can clarify what the spamming scenario here was.

Reply:
Having a random delay between/before requests is some "good" behavior to not dos your own architecture.

Reply:
Addressed by https://github.com/paritytech/cumulus/pull/1891/commits/144445d36f86ecc7e227f2d7161daf4e94ed0028